modules/apps/portal-search-elasticsearch7/portal-search-elasticsearch7-impl/src/main/java/com/liferay/portal/search/elasticsearch7/internal/bootstrap/ElasticsearchBootstrap.java

```diff
@@ -0,0 +1,328 @@
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.util.StringUtil;
+import com.liferay.portal.util.PortalClassPathUtil;
+import com.liferay.portal.util.PropsValues;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+
+import java.net.URI;
+import java.net.URL;
+
+import java.security.CodeSource;
+import java.security.ProtectionDomain;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author Dante Wang
+ */
+public class ElasticsearchBootstrap {
+
+	public ElasticsearchBootstrap(
+			ProcessExecutor processExecutor, String hostname)
+		throws Exception {
+
+		_processExecutor = processExecutor;
+
+		_processConfig = _createProcessConfig(hostname);
+	}
+
+	public void start() throws Exception {
+		_processChannel = _processExecutor.execute(
+			_processConfig,
+			new ElasticsearchBootstrapProcessCallable(new String[0]));
+	}
+
+	public void stop() {
+		if (_processChannel != null) {
+			NoticeableFuture<String> noticeableFuture =
+				_processChannel.getProcessNoticeableFuture();
+
+			noticeableFuture.cancel(true);
+		}
+	}
+
+	private String _createClasspath(File esClasspath) throws Exception {
+		StringBundler sb = new StringBundler();
+
+		// This module
+
+		ProtectionDomain protectionDomain =
+			ElasticsearchBootstrap.class.getProtectionDomain();
+
+		CodeSource codeSource = protectionDomain.getCodeSource();
+
+		URL url = codeSource.getLocation();
+
+		URI uri = url.toURI();
+
+		File file = new File(uri);
+
+		sb.append(file.getAbsolutePath());
+
+		sb.append(File.pathSeparator);
+
+		// $ES_HOME/libs/*
+
+		File[] libFiles = esClasspath.listFiles();
+
+		for (File libFile : libFiles) {
+			sb.append(libFile.getAbsolutePath());
+			sb.append(File.pathSeparator);
+		}
+
+		// Only include petra jars in the classpath.
+		// Elasticsearch has a Jar Hell check, we need to keep classpath minimum
+		// to avoid classes with the same name, even though such classes may not
+		// conflict with Elasticsearch's.
+
+		ProcessConfig portalProcessConfig =
+			PortalClassPathUtil.getPortalProcessConfig();
+
+		PathHolder[] pathHolders =
+			portalProcessConfig.getBootstrapClassPathHolders();
+
+		for (PathHolder pathHolder : pathHolders) {
+			String path = pathHolder.toString();
+
+			if (!path.contains("petra")) {
+				continue;
+			}
+
+			sb.append(path);
+			sb.append(File.pathSeparator);
+		}
+
+		sb.setIndex(sb.index() - 1);
+
+		return sb.toString();
+	}
+
+	private ProcessConfig _createProcessConfig(String hostname)
+		throws Exception {
+
+		// ES ENV: ES_CLASSPATH = $ES_HOME/lib/*
+
+		File esClasspath = new File(_ES_HOME, "lib");
+
+		// ES ENV: ES_PATH_CONF = $ES_HOME/config
+
+		File esPathConf = new File(_ES_HOME, "config");
+
+		// Java executable, use the same one as portal
+
+		String javaExecutable = _getJavaExecutable(
```

Comment:
@dantewang If we simply use "Java -cp" here, it should be the same one as portal, why you have to do the detect?

Tina